### PR TITLE
[WIP] Natural gradient descent, v2

### PR DESCRIPTION
**Pull request by rhaps0dy**
Fork of #1191 

- [x] Add `NaturalVariationalDistribution`
- [x] Add unit/example tests
- [ ] Add tutorial notebook

### CODE REVIEW

File: gpytorch/variational/tril_natural_variational_distribution.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        The :obj:`~gpytorch.variational.NaturalVariationalDistribution`, which
+        needs less iterations to make variational regression converge, at the
+        cost of introducing numerical instability.
+
+    .. note::
+        The relationship of the parameter :math:`\boldsymbol \Theta_\text{tril\_mat}`
+        to the natural parameter :math:`\boldsymbol \Theta_\text{mat}` from
+        :obj:`~gpytorch.variational.NaturalVariationalDistribution` is
+        :math:`\boldsymbol \Theta_\text{mat} = -1/2 {\boldsymbol \Theta_\text{tril\_mat}}^T
+               {\boldsymbol \Theta_\text{tril\_mat}}`.
```
Comment (rhaps0dy) [inline gpytorch/variational/tril_natural_variational_distribution.py:51]: The formatting of lines 50-51 in the readthedocs is all wrong. But `black` splits the line in two so I don't know what to do. Can I nest `.. note` and `.. math`? We'll see.

File: gpytorch/variational/natural_variational_distribution.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+            \\
+            \boldsymbol \Theta_\text{mat} &= -\frac{1}{2} \mathbf S^{-1}.
+        \end{align*}
+
+    The gradients with respect to the variational parameters calculated by this
+    class are instead the **natural** gradients. Thus, optimising its
+    parameters using gradient descent (:obj:`~torch.optim.SGDOptimizer`)
+    becomes natural gradient descent (see e.g. `Salimbeni et al., 2018`_).
+
+    .. note::
```
Comment (rhaps0dy) [inline gpytorch/variational/natural_variational_distribution.py:39]: should be a warning

Comment (gpleiss) [reply gpytorch/variational/natural_variational_distribution.py:39]: Nit: `torch.optim.SGDOptimizer`

File: gpytorch/variational/natural_variational_distribution.py [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    return grad_input.add(grad_input.transpose(-1, -2)).mul_(0.5)
+
+
+class _NaturalToMuVarSqrt(torch.autograd.Function):
+    @staticmethod
+    def _forward(nat_mean, nat_covar):
+        try:
+            L_inv = torch.cholesky(-2.0 * nat_covar, upper=False)
+        except RuntimeError as e:
+            if str(e).startswith("cholesky"):
```
Comment (gpleiss) [inline gpytorch/variational/natural_variational_distribution.py:124]: nit: maybe more future-proof to do `str(e).lower()`?

File: gpytorch/variational/tril_natural_variational_distribution.py (hunk 1 above)
Comment (gpleiss) [reply gpytorch/variational/tril_natural_variational_distribution.py:51]: You could also try replacing `boldsymbol` with `mathbf`? Might save a few characters?

File: gpytorch/variational/natural_variational_distribution.py (hunk 3 above)
Comment (rhaps0dy) [reply gpytorch/variational/natural_variational_distribution.py:124]: The Pytorch developers should have just raised a `numpy.LinAlgError` :'(

File: gpytorch/variational/natural_variational_distribution.py (hunk 2 above)
Comment (rhaps0dy) [reply gpytorch/variational/natural_variational_distribution.py:39]: What's the difference? I'm confused.

Comment (gpleiss) [reply gpytorch/variational/natural_variational_distribution.py:39]: agreed

Comment (gpleiss) [reply gpytorch/variational/natural_variational_distribution.py:39]: line 41: there's a typo

Comment (rhaps0dy) [reply gpytorch/variational/natural_variational_distribution.py:39]: Ah sorry I was looking at line 36
